Add row group customization

edenhaus opened the pull request:
Add the possibility to change the row group of a table similar to hide/view columns. The row group configuration will be stored with the other state settings of the table in the user settings.

Screenshot of the shoppinglist customizations:
![grafik](https://user-images.githubusercontent.com/26537646/102270113-e7c54a00-3f1d-11eb-84c4-d02974c25e19.png)

Selecting "none" will disable the rowGroup feature and disable the fixed order.
Selecting a column will group it after the column and also activate a fixed order for that column. (In the code I have found, that when rowGroup is activated also fixedOrder on that column was applied)


Currently the Group by section is shown only if rowGroup extension is added to table, which is currently the case for shoppinglist, tasks, productform, recipeform.
@berrnd Should we enable the possibility for all tables?

edenhaus commented:
After rethinking what is the best translation of "None" in German, I think maybe we should rename it to "Deactivated" or what you think?

berrnd commented:
Cool, thanks - looks good from a first sight. :)

Will give it a short test tomorrow or so - and I think "None" is fine and it would be also ok to enable this for all tables (some columns will not make sense to group by them, but I think we don't need to limit that).

With more table options your other request #1168 is probably getting more important and maybe also a "Reset" button.

edenhaus commented:
I will add it for all tables this afternoon :)

For the Reset button there exist a method `dataTable.state.clear()` but I need to give it a closer look.

edenhaus commented:
ok there is still a bug somewhere. I was trying to add it for all tables and found out, that after the loadStateEvent a saveStateEvent is fired, where the rowGroup is set to the default value.

Example for reproduction:
- Open Shoppinglist and set row group to Amount
- Refresh the page and the row group is setting back to Product group

edenhaus commented:
Took me some time to get a working solution. RowGroup unfortunately cannot be managed via the state events, but I have implemented via the preDrawCallback.

Also I have activated rowGroup for all tables and also implemented a reset button.
![grafik](https://user-images.githubusercontent.com/26537646/102378073-b0aa7380-3fc5-11eb-87a2-b1a3b3060342.png)

After clicking on reset a confirmation window will be open. Confirming will delete the usersetting entries in the db and reload the page. The later one is required as `state.clear()` is only calling `stateSaveCallback` but not refreshing the table afterwards.
![grafik](https://user-images.githubusercontent.com/26537646/102378473-24e51700-3fc6-11eb-829e-acf2cb3cade4.png)

I have tested it with some tables without any problems.

berrnd reviewed (approved):
Works, thanks a lot! :)

> How is the openapi file generated? Manually?

Yes, I've added the new endpoint `DELETE /user/settings/{settingKey}`.

berrnd commented:
I've also added that row groups are collapsible (like we already had it for the shopping list) for alle tables in https://github.com/grocy/grocy/commit/cc2ea93313add269d7f36cce0ae404cc7fc20128. 🙂
